src/PlatformInstaller/DotNetUpgrade/RuntimeUpgradeManager.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.ComponentModel;
+using System.IO;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Anotar.Serilog;
+using Caliburn.Micro;
+using Microsoft.Win32;
+
+public class RuntimeUpgradeManager
+{
+    IEventAggregator eventAggregator;
+    CredentialStore credentialStore;
+    ProcessRunner processRunner;
+    const string dotNet452WebInstallerExe = "NDP452-KB2901954-Web.exe";
+
+    // This download link came from .NET Framework Deployment Guide for Developers -  see https://msdn.microsoft.com/en-us/library/ee942965%28v=vs.110%29.aspx#redist
+    const string dotNet452WebInstallerURL = @"http://go.microsoft.com/fwlink/?LinkId=397707";
+    FileInfo installer;
+
+    public RuntimeUpgradeManager(IEventAggregator eventAggregator, CredentialStore credentialStore, ProcessRunner processRunner)
+    {
+        this.eventAggregator = eventAggregator;
+        this.credentialStore = credentialStore;
+        this.processRunner = processRunner;
+
+        var tempFolder = new DirectoryInfo(Environment.ExpandEnvironmentVariables("%temp%"));
+        var assetFolder = Directory.CreateDirectory(Path.Combine(tempFolder.FullName, @"Particular\PlatformInstaller"));
+        installer = new FileInfo(Path.Combine(assetFolder.FullName, dotNet452WebInstallerExe));
+    }
+    
+    public bool Is452orLaterInstalled()
+    {
+        using (var ndpKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32).OpenSubKey(@"SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full\"))
+        {
+            if (ndpKey?.GetValueKind("Release") == RegistryValueKind.DWord)
+            {
+                var release = (int) ndpKey.GetValue("Release");
+                return release >= 379893;
+            }
+        }
+        return false;
+    }
+
+    public async Task<FileInfo> Download452WebInstaller()
+    {
+        eventAggregator.PublishOnUIThread(new DotNetDownloadStartedEvent());
+        using (var client = new WebClient())
+        {
+            client.Proxy.Credentials = credentialStore.Credentials;
+            client.DownloadProgressChanged += OnClientOnDownloadProgressChanged;
+            client.DownloadFileCompleted += OnClientOnDownloadFileCompleted;
+
+            var retries = 0;
+            const int maxretries = 5;
+
+            while (true)
+            {
+                LogTo.Information("Attempting to download '{0}' to '{1}'", dotNet452WebInstallerURL, installer.FullName);
+                try
+                {
+                    await client.DownloadFileTaskAsync(dotNet452WebInstallerURL, installer.FullName).ConfigureAwait(false);
+                    return installer;
+                }
+                catch
+                {
+                    retries++;
+                    PublishFailed();
+                    Thread.Sleep(500);
```

Comment:
i think inside an async  this is better

    await Task.Delay(500).ConfigureAwait(false);